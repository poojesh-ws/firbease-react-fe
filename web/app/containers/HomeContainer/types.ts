import { AnyAction } from 'redux';
import { IntlShape } from 'react-intl';
import { GqlQueryReponse } from '@app/utils/graphqlUtils';
export interface Launch {
  id: string;
  missionName: string;
  launchDateUtc: string;
  launchDateUnix: number;
  links: {
    wikipedia: string;
    flickrImages: Array<string>;
  };
}

export interface LaunchData {
  launches?: Launch[];
}
export interface RequestLaunchesActionPayload {
  missionName: string | null;
  order: string | null; // 'asc' | 'desc'
  page: number; // starts from 1
}

export type LaunchesActionCreator = (payload: RequestLaunchesActionPayload) => AnyAction;

export interface HomeContainerProps {
  dispatchLaunchList: LaunchesActionCreator;
  launchData: LaunchData;
  launchListError?: string;
  loading: boolean;
  intl: IntlShape;
}

export type LaunchesResponse = GqlQueryReponse<{ launches?: Launch[] }>;

export interface LaunchesAction {
  payload: { missionName: any; order: any; page: any };
  type: string;
}
