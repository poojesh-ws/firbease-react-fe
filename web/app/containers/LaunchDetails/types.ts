import { AnyAction } from 'redux';
import { GqlQueryReponse } from '@app/utils/graphqlUtils';
export interface LaunchDetails {
  id: string;
  missionName: string;
  details: string;
  rocket: {
    rocketName: string;
    rocketType: string;
  };
  ships: {
    name: string;
    type: string;
  }[];
  links: {
    flickrImages?: string[];
  };
}

export type LaunchResponse = GqlQueryReponse<LaunchDetails>;

export interface LaunchDetailsProps {
  launch: LaunchDetails | null;
  launchError?: string;
  loading: boolean;
  dispatchLaunch: (launchId: string) => AnyAction;
}
