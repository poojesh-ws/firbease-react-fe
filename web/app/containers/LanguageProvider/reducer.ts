import { createSlice } from '@reduxjs/toolkit';
import { DEFAULT_LOCALE } from '@app/i18n';

export const initialState = {
  locale: DEFAULT_LOCALE
};

const languageProviderSlice = createSlice({
  name: 'language',
  initialState,
  reducers: {
    changeLocale(state, action) {
      state.locale = action.payload;
    }
  }
});

export const { changeLocale } = languageProviderSlice.actions;

export default languageProviderSlice.reducer;
