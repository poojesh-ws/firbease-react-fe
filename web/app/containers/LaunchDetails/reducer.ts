import { createSlice } from '@reduxjs/toolkit';
import get from 'lodash-es/get';
import { prepare } from '@app/utils';

export const initialState = {
  launch: {},
  loading: true,
  launchError: null
};

const launchDetailsSlice = createSlice({
  name: 'launchDetails',
  initialState,
  reducers: {
    requestGetLaunch: {
      reducer: (state) => {
        state.loading = true;
      },
      prepare
    },
    successGetLaunch(state, action) {
      state.loading = false;
      state.launchError = null;
      state.launch = get(action.payload, 'launch', {});
    },
    failureGetLaunch(state, action) {
      state.loading = false;
      state.launch = {};
      state.launchError = get(action.payload, 'message', 'something_went_wrong');
    }
  }
});

export const { requestGetLaunch, successGetLaunch, failureGetLaunch } = launchDetailsSlice.actions;

export default launchDetailsSlice.reducer;
