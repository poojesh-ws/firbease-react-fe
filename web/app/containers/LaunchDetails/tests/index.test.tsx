/**
 *
 * Tests for LaunchDetails
 *
 *
 */

import React from 'react';
import { renderProvider, timeout } from '@utils/testUtils';
// import { fireEvent } from '@testing-library/dom'
import { LaunchDetailsTest as LaunchDetails, mapDispatchToProps } from '..';
import { LaunchDetailsProps } from '../types';
import history from '@app/utils/history';
import { requestGetLaunch } from '../reducer';

describe('<LaunchDetails /> container tests', () => {
  let submitSpy: jest.Mock;
  let defaultProps: LaunchDetailsProps;

  beforeEach(() => {
    submitSpy = jest.fn();
    defaultProps = {
      loading: false,
      launch: null,
      dispatchLaunch: submitSpy
    };
  });
  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<LaunchDetails {...defaultProps} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should call dispatchLaunch if launchId param is present', async () => {
    history.location.pathname = '/1';
    renderProvider(<LaunchDetails {...defaultProps} />, { path: '/:launchId' });
    await timeout(500);
    expect(submitSpy).toBeCalledWith('1');
  });

  it('should mapDispatchToProps works as expected', () => {
    const dispatchSpy = jest.fn();
    const props = mapDispatchToProps(dispatchSpy);
    props.dispatchLaunch('1');
    expect(dispatchSpy).toBeCalledWith(requestGetLaunch('1'));
  });
  it('should not render the launchDetails if there is no data', () => {
    const { getByTestId } = renderProvider(<LaunchDetails {...defaultProps} />);
    expect(() => getByTestId('launch-details')).toThrowError();
  });
  it('should  render the launchDetails if there is data', () => {
    const props = {
      loading: false,
      launch: {
        id: '1',
        missionName: 'CRS-21',
        links: {
          flickrImages: ['https://farm9.staticflickr.com/8617/16789019815_f99a165dc5_o.jpg']
        },
        details: "SpaceX's 21st ISS resupply mission.",
        rocket: {
          rocketName: 'Falcon 9',
          rocketType: 'FT'
        },
        ships: [
          {
            name: 'Ship 1',
            type: 'Type 1'
          }
        ]
      },
      dispatchLaunch: submitSpy
    };

    const { getByTestId } = renderProvider(<LaunchDetails {...props} />);

    expect(getByTestId('launch-details')).toBeInTheDocument();
  });
});
