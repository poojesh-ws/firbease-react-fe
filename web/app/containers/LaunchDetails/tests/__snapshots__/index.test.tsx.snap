// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LaunchDetails /> container tests should render and match the snapshot 1`] = `
<body>
  <div>
    <div>
      <div
        class="NotFoundPage__NotFoundContainer-sc-13pqvdn-1 hNyuuR"
      >
        <p
          class="T__StyledText-znbtqz-0 cJIIqm"
          data-testid="details"
        >
          Oops you ran out of Oxygen!, dont worry, you can go back home and start exploring again!
        </p>
        <button
          class="NotFoundPage__CustomButton-sc-13pqvdn-2 llAsg"
          data-testid="back-button"
        >
          Go Back
        </button>
        <div
          class="ant-image"
        >
          <img
            class="ant-image-img NotFoundPage__NotFoundImage-sc-13pqvdn-0 dtcQqo"
            src="IMAGE_MOCK"
          />
        </div>
      </div>
      <p
        class="T__StyledText-znbtqz-0 hGCwnP"
        data-testid="default-message"
      />
    </div>
  </div>
</body>
`;
